capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -212,8 +213,14 @@ static VolumeData individualBookToVolumeData(JsonElement singleBook) {
     String canonicalVolumeLink = canonVolumeElement != null ? canonVolumeElement.getAsString() : "";
 
     JsonElement thumbnailElement = volumeInfoObj.get("imageLinks");
-    String thumbnailLink = thumbnailElement != null ? thumbnailElement.getAsJsonObject().get("thumbnail").getAsString()
-        : "";
+    ArrayList<String> sizes = new ArrayList<>(Arrays.asList("extraLarge", "large", "medium", "small", "thumbnail"));
+
+    String thumbnailLink = "";
+    String greatestSize = sizes.stream().filter(size -> thumbnailElement.getAsJsonObject().get(size) != null)
+        .findFirst().get();
```

Comment:
Note that calling `Optional.get` here will throw a `NoSuchElementException` if none of the sizes make it past the previous `filter` step. The fact that the `"thumbnail"` element is guaranteed to be there should ensure this won't happen, but I think it might be a good idea to program defensively here. For example, instead of calling `.get` here, you could instead add an `ifPresent` check to the conditional below.

Reply:
Fixed! I didn't know about the Optional data type in Java but that is interesting!